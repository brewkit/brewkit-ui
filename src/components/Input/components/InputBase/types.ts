import { ReactNode } from 'react';

export interface Props {

    /**
     * The component used for the `input` element
     */
    as?: 'input' | 'select' | 'textarea',

    /**
     * Whether or not the radio or checkbox input is checked
     *
     * @default false
     */
    checked?: boolean,

    /**
     * The `children` prop can be used for certain inputs like `select` and `textarea`
     */
    children?: ReactNode,

    /**
     * The wrapper class
     */
    className?: string,

    /**
     * The number of columns a textarea has.
     */
    cols?: number,

    /**
     * The custom control to show instead of the default control.  Can be used for any input type,
     * but is intended mostly for inputs that don't accept styling well like select boxes, radio buttons and
     * checkboxes.
     */
    customControl?: ReactNode,

    /**
     * Whether the `input` is checked by default.  Used for checkboxes.
     */
    defaultChecked?: boolean,

    /**
     * The `defaultValue` of the `input` element
     *
     * @default undefined
     */
    defaultValue?: any,

    /**
     * If `true`, the `input` will be disabled
     */
    disabled?: boolean,

    /**
     * If `true`, the input will apply error styling
     */
    error?: boolean,

    /**
     * If `true`, the input will be 100% of it's container width
     */
    fullWidth?: boolean,

    /**
     * The `input` element id
     */
    id?: string,

    /**
     * The `input` element className.
     */
    inputClassName?: string,

    /**
     * The `input` label for radio buttons and checkboxes
     */
    inputLabel?: string,

    /**
     * The className for the input label
     */
    inputLabelClassName?: string,

    /**
     * Whether or not the `select` input accepts multiple values
     */
    isMultiple?: boolean,

    /**
     * The `input` element name
     */
    name?: string,

    /**
     * The `input` element onChange handler
     */
    onChange?: any,

    /**
     *  The HTML5 placeholder applied to the `input` element
     */
    placeholder?: string,

    /**
     * Use this to define a read-only field
     */
    readOnly?: boolean,

    /**
     * If `true`, the `input` element will be required
     */
    required?: boolean,

    /**
     * The number of rows a textarea has.
     */
    rows?: number,

    /**
     * The `input` element type.  Can be any valid HTML5 input type
     */
    type?: 'button'
    | 'checkbox'
    | 'color'
    | 'date'
    | 'datetime-local'
    | 'email'
    | 'file'
    | 'hidden'
    | 'image'
    | 'month'
    | 'number'
    | 'password'
    | 'radio'
    | 'range'
    | 'reset'
    | 'search'
    | 'submit'
    | 'tel'
    | 'text'
    | 'time'
    | 'url'
    | 'week',

    /**
     *  The `value` of the `input` element
     */
    value?: any,

    /**
     * Content that can be injected after the Input
     */
    after?: ReactNode,

    /**
     * Content that can be injected before the Input
     */
    before?: ReactNode,

}
