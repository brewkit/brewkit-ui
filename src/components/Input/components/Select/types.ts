import { Props as InputBaseProps } from '../InputBase/types';

export type Props = Omit<InputBaseProps,
    'as'
    | 'after'
    | 'before'
    | 'checked'
    | 'cols'
    | 'customControl'
    | 'defaultChecked'
    | 'inputLabel'
    | 'inputLabelClassName'
    | 'rows'
    | 'type'
>;
