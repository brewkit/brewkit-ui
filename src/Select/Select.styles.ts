import { createStyles, makeStyles } from '@material-ui/core';

export default makeStyles(({ palette, sizes }) =>
    createStyles({
        root: {
            '&$select': {
                transition: 'all 0.3s',
            },
            '&$select:focus': {
                background: palette.primary.main,
                color: palette.common.white,
                borderRadius: '4px',
            },
            '&:focus ~ svg': {
                color: palette.common.white,
            },
            '&$select option': {
                color: palette.background.default,
                padding: sizes.small,
            },
            '& label': {
                transform: 'translate(18px, 18px) scale(1)',
            },
        },
        iconOpen: {
            color: palette.common.white,
        },

        menu: {
            padding: sizes.xsmall,
            backgroundColor: palette.common.white,
            border: `1px solid ${palette.primary.main}`,

            '& ul': {
                padding: 0,
            },
        },

        menuItem: {
            padding: sizes.small,
            transition: 'all 0.3s',
        },

        activeMenuItem: {
            backgroundColor: `${palette.primary.main} !important`,
            color: palette.common.white,
        },
    }),
);
