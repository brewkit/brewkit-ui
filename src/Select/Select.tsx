import React from 'react';
import {
    Select as MUISelect,
    SelectClassKey,
    SelectProps as MUISelectProps,
} from '@material-ui/core';
import useStyles from './Select.styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import { merge } from 'lodash';

type BkSelectProps = MUISelectProps & {
    classes?: ClassNameMap<
        SelectClassKey | 'menu' | 'menuItem' | 'activeMenuItem'
    >;
};

const Select = React.forwardRef(
    (props: BkSelectProps, ref: React.Ref<HTMLSelectElement>) => {
        const {
            classes: userClasses = {} as BkSelectProps['classes'],
            ...otherProps
        } = props;
        const classes = merge(useStyles(), userClasses);

        return <MUISelect classes={classes} ref={ref} {...otherProps} />;
    },
);

Select.displayName = 'Select';

export default Select;
