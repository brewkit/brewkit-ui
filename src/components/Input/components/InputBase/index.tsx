import React, { ReactElement } from 'react';
import clsx from 'clsx';
import Typography from '../../../Typography';
import { Props } from './types';

/**
 * This component is made to be extended, not to be used directly
 */
const InputBase = ({
    after,
    as = 'input',
    before,
    checked,
    children,
    className,
    cols,
    customControl,
    defaultChecked,
    defaultValue,
    disabled,
    error,
    fullWidth,
    id,
    inputClassName,
    inputLabel,
    inputLabelClassName,
    isMultiple,
    name,
    onChange,
    placeholder,
    readOnly,
    required,
    rows,
    type = 'text',
    value,
    ...otherProps
}: Props): ReactElement => {
    const classes = clsx(
        'brew-Input',
        { 'brew-Input--isDisabled': Boolean(disabled) },
        { 'brew-Input--hasError': Boolean(error) },
        { 'brew-Input--isFullWidth': Boolean(fullWidth) },
        { 'brew-Input--hasBefore': Boolean(before) },
        { 'brew-Input--hasAfter': Boolean(after) || (!disabled && error) },
        className,
    );

    const Element = as;

    return (
        <Typography as="label" className={classes} {...otherProps}>
            {before}

            <Element
                checked={checked}
                className={clsx('brew-Input__element', inputClassName)}
                cols={cols}
                defaultChecked={defaultChecked}
                defaultValue={defaultValue}
                disabled={disabled}
                id={id}
                multiple={isMultiple}
                name={name}
                onChange={onChange}
                placeholder={placeholder}
                readOnly={readOnly}
                required={required}
                rows={rows}
                type={type}
                value={value}
            >
                {children}
            </Element>

            {after}

            {Boolean(customControl) && customControl}

            {inputLabel && (type === 'checkbox' || type === 'radio') && (
                <span className={clsx('brew-Input__label', inputLabelClassName)}>{inputLabel}</span>
            )}
        </Typography>
    );
};

export default InputBase;
