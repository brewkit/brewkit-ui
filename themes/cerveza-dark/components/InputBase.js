import { color } from '../../cerveza/variables';
var InputBase = {
    overrides: {
        MuiInputBase: {
            '&$disabled': {
                color: color.gray.lightest
            }
        }
    }
};
export default InputBase;
