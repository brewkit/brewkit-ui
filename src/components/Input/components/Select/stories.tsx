import React, { ReactElement } from 'react';
import Select from './index';


export default {
    component: Select,
    title: 'Components|User Input/Input/Select',
};


export const Default = (): ReactElement => (


    <Select defaultValue="2">
        <option value="1">option 1</option>
        <option disabled value="2">option 2</option>
        <option value="3">option 3</option>
    </Select>


);


export const Overflow = (): ReactElement => (


    <Select value="option 2">
        <option value="1">option 1</option>
        <option value="2">option 2</option>
        <option value="3">option 3</option>
        <option value="4">option 4</option>
        <option value="5">option 5</option>
        <option value="6">option 6</option>
    </Select>


);
