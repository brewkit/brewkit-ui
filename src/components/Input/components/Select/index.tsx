/* eslint-disable max-statements */

import React, { ReactElement } from 'react';
import clsx from 'clsx';
import InputBase from '../InputBase';
import { Props } from '../InputBase/types';
import DefaultCustomControl from './components/DefaultCustomControl';


/**
 * The select component is the brewkit implementation of the HTML select, with a few
 * extra features.
 */
const Select = ({
    className,
    children,
    onChange: onChangeProp,
    value: valueProp,
    defaultValue: defaultValueProp,
    ...otherProps
}: Props): ReactElement => {

    if (
        !React.Children.toArray(children).every((child: any) => child?.type === 'option')
    ) throw new Error('All children of a select element must be options');

    // Since we need to set the value of the select input, we can't allow it to be uncontrolled
    const [value, setValue] = React.useState(valueProp || defaultValueProp || '');

    // It would help for us to know what the text was of the input, not just the value
    const [text, setText] = React.useState('');

    // If it's a controlled component, we need to keep up with the value
    React.useEffect(() => {
        if (valueProp || valueProp === '') setValue(valueProp);
    }, [valueProp]);

    // Get the text of the option based on the value
    React.useEffect(() => {
        const textVal = React.Children
            .toArray(children)
            .reduce((
                curr: string,
                option: any,
                index: number,
                arr: any[],
            ) => { // eslint-disable-line max-params

                const { value: optionValue, children: optionChildren } = option?.props || {};
                if (optionValue === value ||
                (!optionValue && optionValue !== '' && optionChildren === value)) return optionChildren;
                if (!curr && index === arr.length - 1) return arr[0].props.children;

                return curr;
            }, '');
        setText(textVal);
    }, [value]);

    // pass a custom onChange event
    const onChange = (event: React.ChangeEvent<HTMLSelectElement>): void => {
        const { value: inputValue } = event.target;
        setValue(inputValue);
        if (onChangeProp) onChangeProp(event);
    };

    // handle the event where someone clicks one of our "fake" custom control options
    const handleOptionClick = (optionValue: string): void => {
        setValue(optionValue);
    };

    const classes = clsx(
        'brew-Select',
        className,
    );

    return (
        <InputBase
            as="select"
            className={classes}
            customControl={(
                <DefaultCustomControl
                    inputChildren={children}
                    onSelectOption={handleOptionClick}
                    text={text}
                />)}
            onChange={onChange}
            value={value}
            {...otherProps}
        >
            {children}
        </InputBase>
    );


};


export default Select;
