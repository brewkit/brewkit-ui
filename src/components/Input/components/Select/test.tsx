import React, { ReactElement } from 'react';
import renderer from 'react-test-renderer';
import Select from './index';
import InputBase from '../InputBase';
import { shallow, mount } from 'enzyme';


describe('Select', () => {

    
    it('renders correctly', () => {
        const tree = renderer
            .create(<Select name="name"><option>Option 1</option><option>Option 2</option></Select>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('applies the correct base classes', () => {
        const select = shallow(<Select />);
        expect(select.hasClass('brew-Select')).toBe(true);
    });

    it('uses the InputBase component', () => {
        const select = shallow(<Select />);
        expect(select.matchesElement(<InputBase />)).toBe(true);
    });

    it('renders a `select` element', () => {
        const select = shallow(<Select />);
        expect(select.render().find('select').length).toBe(1);
    });

    it('accepts options as children', () => {
        const select = shallow(<Select><option>an option</option></Select>);
        expect(select.render().find('option').length).toBe(1);
    });

    it('correctly passes through a classname', () => {
        const testClassName = `class-${Math.floor(Math.random() * 1000)}`;
        const select = shallow(<Select className={testClassName} />);
        expect(select.hasClass(testClassName)).toBe(true);
    });

    it('passes through an onchange handler', () => {
        const onChangeMock = jest.fn();
        const select = mount(<Select onChange={onChangeMock}/>);
        select.find('select').simulate('change');
        expect(onChangeMock).toHaveBeenCalled();
    });


});

