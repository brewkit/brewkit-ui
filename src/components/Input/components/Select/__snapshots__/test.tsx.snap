// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly 1`] = `
<label
  className="brew-Typography brew-Typography--variant-inherit brew-Typography--color-initial brew-Typography--align-inherit brew-Typography--display-initial brew-Input brew-Select"
>
  <select
    className="brew-Input__element"
    name="name"
    onChange={[Function]}
    type="text"
    value=""
  >
    <option>
      Option 1
    </option>
    <option>
      Option 2
    </option>
  </select>
  <div
    className="brew-CustomControl"
  >
    <div
      className="brew-CustomControl__input"
      onClick={[Function]}
    >
      <div
        className="brew-CustomControl__renderedSelection"
      >
        
      </div>
      <i
        className="brew-Icon material-icons brew-Icon--color-inherit brew-Icon--size-inherit brew-CustomControl__icon"
      >
        keyboard_arrow_down
      </i>
    </div>
  </div>
</label>
`;
